src/hexer/duplifier.nim

```diff
@@ -161,12 +161,59 @@ proc isResultUsage(c: Context; n: Cursor): bool {.inline.} =
   if n.kind == Symbol:
     result = n.symId == c.resultSym
 
+proc isSimpleExpression(n: var Cursor): bool =
+  ## expressions that can be returned safely
+  case n.kind
+  of Symbol, UIntLit, StringLit, IntLit, FloatLit, CharLit, DotToken, Ident:
+    result = true
+    inc n
+  of ParLe:
+    case n.exprKind
+    of FalseX, TrueX, InfX, NegInfX, NanX, NilX, SufX:
+      result = true
+      skip n
+    of CastX, ConvX, HconvX, DconvX:
+      result = true
+      inc n
+      skip n # type
+      while n.kind != ParRi:
+        if not isSimpleExpression(n): return false
+      skipParRi n
+    of ExprX:
+      inc n
+      var inner = n
+      skip n
+      if n.kind == ParRi:
+        result = isSimpleExpression(inner)
+        skipParRi n
+      else:
+        result = false
+    else:
+      result = false
+      skip n
+  of ParRi, SymbolDef, UnknownToken, EofToken:
+    result = false
+    inc n
+
 proc trReturn(c: var Context; n: var Cursor) =
-  copyInto c.dest, n:
+  var ret = createTokenBuf()
```

Comment:
Instead of the `ret` temporary and the swap dance, use this:

```nim
  let retVal = n.firstSon
  var exp = retVal
  if isResultUsage(c, retVal):
    copyInto(c.dest, n): takeTree c, n
  elif isSimpleExpression(exp) or c.resultSym == NoSymId:
    copyInto(c.dest, n): tr(c, n, WantOwner)
 else:
   ...

```